go_lib/deckhouse-config/validator.go

```diff
@@ -46,74 +46,139 @@ type ValuesValidator interface {
+func (c *ConfigValidator) ValidateCR(cfg *d8cfg_v1alpha1.ModuleConfig) ValidationResult {
 	result := ValidationResult{}
 
-	// It is ok if resource spec is empty or if it has enabled field only.
-	if len(cfg.Spec.Settings) == 0 && cfg.Spec.Version == 0 {
-		return result, nil
+	if cfg.Spec.Version == 0 {
+		// Resource is not valid when spec.settings are specified without version.
+		if cfg.Spec.Settings != nil {
+			result.ValidateCRError = "spec.version is required when spec.settings are specified"
+		}
+		// Resource is valid without spec.version and spec.settings.
+		return result
 	}
-	if cfg.Spec.Settings == nil && cfg.Spec.Version > 0 {
-		return result, fmt.Errorf("spec.version is forbidden when settings are not specified")
+
+	// Can run conversions and validations if spec.version and spec.settings are specified.
+	if cfg.Spec.Settings == nil {
+		// Warn about spec.version without spec.settings.
+		result.ValidateCRWarning = "spec.version is redundant without spec.settings"
+	} else {
+		// Resource has both spec.settings and spec.version — it is ok to convert and validate with OpenAPI schema.
+		result.HasVersionedSettings = true
 	}
 
-	// Settings are present, check version using conversion chain.
+	// Check if there is registered conversion for the version and if the version is the latest.
 	chain := conversion.Registry().Chain(cfg.GetName())
 	latestVer := chain.LatestVersion()
 
+	// Check if version is unknown.
 	if !chain.IsKnownVersion(cfg.Spec.Version) {
 		previousVersions := concatIntList(chain.PreviousVersionsList())
 		if previousVersions != "" {
 			previousVersions = fmt.Sprintf(", or one of previous versions: %s", previousVersions)
 		}
 
-		return result, fmt.Errorf("spec.version=%d is invalid. Use latest version %d%s", cfg.Spec.Version, latestVer, previousVersions)
+		msg := fmt.Sprintf("spec.version=%d is unsupported. Use latest version %d%s", cfg.Spec.Version, latestVer, previousVersions)
+		if result.HasVersionedSettings {
+			// Error if spec.settings are specified. Can't start conversions for such configuration.
+			result.ValidateCRError = msg
+		} else {
+			// Warning if there are no spec.settings.
+			result.ValidateCRWarning = msg
+		}
+		return result
 	}
 
+	// Warning if version is not the latest.
 	versionMsg := ""
 	if cfg.Spec.Version != latestVer {
-		versionMsg = fmt.Sprintf("spec.version=%d is obsolete. Please migrate settings to the latest version %d", cfg.Spec.Version, latestVer)
+		versionMsg = fmt.Sprintf("spec.version=%d is obsolete. Please migrate spec.settings to the latest version %d", cfg.Spec.Version, latestVer)
+	}
+	result.ValidateCRWarning = versionMsg
+	return result
+}
+
+// ConvertToLatest checks if ModuleConfig resource is well-formed and runs conversions for spec.settings is needed.
+func (c *ConfigValidator) ConvertToLatest(cfg *d8cfg_v1alpha1.ModuleConfig) ValidationResult {
+	result := c.ValidateCR(cfg)
+	if result.HasError() || !result.HasVersionedSettings {
+		return result
```

Comment:
Nitpicking

This code would be cleaner this way:

```go
if cfg.Settings == nil {
  return nil 
}
err := := c.ValidateCR(cfg)
if err != nil {
  return err
}
```

Seems, the ValidationResult copies the state of MC and you now track two states of the same data (MC + ValidationResult) instead of one. But what you are trying to do is to get an error, and the knowledge about non-empty settings and changed version. The latter two are easily accessible without ValidationResult. 

If we are to throw away ValidationResult, the code will remain responsible to combine multiple errors into one. [multierr](https://pkg.go.dev/github.com/uber-go/multierr#Combine) package must be useful here.

Additionally, we can wrap errors if we need to split the processing of them. But seems is it not the case.

Additionally, I think currently it is not idiomatic go code.

Reply:
See https://github.com/deckhouse/deckhouse/pull/3125#discussion_r1032373001